doc: Document compiler configuration for native depends packages

willcl-ark opened the pull request:
Fixes: #33859

Previously one had to read the Makefile (and various *.mk configuration
files) to see how to correctly override CC and CXX when building native
depends packages.

Detail this in README.md to make it clearer.

hebasto reviewed (commented):
Concept ACK. Thanks!

hebasto commented on depends/README.md:
````diff
@@ -108,6 +108,58 @@ The following can be set when running make: `make FOO=bar`
+To ensure both native and target use Clang:
+
+```
+make build_CC=clang build_CXX=clang++ host_CC=clang host_CXX=clang++
````
How is this better than `make build_CC=clang build_CXX=clang++ CC=clang CXX=clang++`?

sedited replied:
fwiw I don't recall ever setting `host_CC` in lieu of `CC`, nor having the expectation that this was supported. I concur with hebasto, and don't think they should be documented. The patch you suggest @willcl-ark seems fine in isolation, but since we also support native builds through depends, I think I would still prefer to just document that setting `CC` is supported.

maflcko reviewed (approved):
lgtm 

review ACK 17577646f77e20783ffdd9f322f85e96da2265 🌴

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 17577646f77e20783ffdd9f322f85e96da2265 🌴
b6tuB1ff+N8xGzUilBsW1CSxIOpk4gK7SE473C8JMS+vg4aU5H4YbP04c0HL3NEDxTZkxKFJBRcJ9RVAQ9t7DQ==
```

</details>

hebasto commented on depends/README.md:
````diff
@@ -8,66 +8,96 @@ as well as [packages.md](packages.md) for how to add packages.
 
-    apt install bison g++ ninja-build pkgconf python3 xz-utils
+```
+apt install bison g++ ninja-build pkgconf python3 xz-utils
````
Other native package, not only `native_qt`, require `g++`.

hebasto reviewed (approved):
ACK b1b559f83f629c7d0976c05d07cbc6ec04e84095.

willcl-ark commented on depends/README.md:
```diff
@@ -108,15 +140,67 @@ The following can be set when running make: `make FOO=bar`
+These control compilers used to compile tools that run natively on the build machine during compilation:
+
+- `build_CC`: C compiler for native tools (default: `gcc` on Linux, `clang` on macOS/BSD)
+- `build_CXX`: C++ compiler for native tools (default: `g++` on Linux, `clang++` on macOS/BSD)
```
Taken in 75155c073d7

hebasto reviewed (approved):
re-ACK 7dd714ae71fd18eda82ab4b43d4cecc047b87a2d.

fanquake reviewed (changes requested):
See https://github.com/bitcoin/bitcoin/pull/33902#discussion_r2549756641.

ryanofsky reviewed (approved):
Code review ACK 7dd714ae71fd18eda82ab4b43d4cecc047b87a2d. Nice documentation improvement.

I'm a little confused about hebasto's comments https://github.com/bitcoin/bitcoin/pull/33902#discussion_r2561967012 and https://github.com/bitcoin/bitcoin/pull/33902#issuecomment-3603896660 which seem to say we should not recommend setting variables like `host_CC` and `host_CXX` because setting `CC` and `CXX` does the same thing and these are more standard. If so, would it be better for `README.md` to remove the `host_` prefixes? I assume current version is ok given the ACKs, but this seems like something that would be good to follow up on or at least respond to.

Also fanquake's comment https://github.com/bitcoin/bitcoin/pull/33902#discussion_r2549756641 seems to say that setting CC/CXX works better than host_CC/host_CXX in practice? So maybe the readme should be updated to just drop the `host_` prefix entirely if the `host_` prefixed variables are less standard and don't work as well as the unprefixed variables.

hebasto commented:
> I'm a little confused about hebasto's comments [#33902 (comment)](https://github.com/bitcoin/bitcoin/pull/33902#discussion_r2561967012) and [#33902 (comment)](https://github.com/bitcoin/bitcoin/pull/33902#issuecomment-3603896660) which seem to say we should not recommend setting variables like `host_CC` and `host_CXX` because setting `CC` and `CXX` does the same thing and these are more standard.

Yes. That's my point.

> Also fanquake's comment [#33902 (comment)](https://github.com/bitcoin/bitcoin/pull/33902#discussion_r2549756641) seems to say that setting CC/CXX works better than host_CC/host_CXX in practice? So maybe the readme should be updated to just drop the `host_` prefix entirely if the `host_` prefixed variables are less standard and don't work as well as the unprefixed variables.

I'll support such changes once someone proposes them.

willcl-ark committed:
doc: document depends compiler configuration

CC/CXX only override compilers for target (host) packages. Native
build tool packages (native_capnp, native_qt, etc.) use separate
build_CC/build_CXX variables, which default to gcc/g++ on Linux.

On systems without gcc (e.g. Nix, Chimera Linux), native package
builds fail unless build_CC/build_CXX are also set explicitly.
Document how to override both sets of compilers.

hebasto reviewed (approved):
ACK 4c07cf87e2288d4a8b0174ad64e52f455de9029e.

sedited reviewed (approved):
ACK 4c07cf87e2288d4a8b0174ad64e52f455de9029e

fanquake reviewed (approved):
ACK 4c07cf87e2288d4a8b0174ad64e52f455de9029e
